### ISSUE TITLE
[Proposal] Eloquent doesn't need to execute relation query if foreign key is null

### ISSUE THREAD

**Issue by cviebrock** [laravel/ideas#748]
(Originally submitted as a bug here: https://github.com/laravel/framework/issues/20616)

### Proposal:

When the foreign key of a relation is `null`, then don't execute the SQL call when trying to load the relation.

### Example:

Assume the following Eloquent model that has an (optional) relationship (i.e. the `author_id` field is nullable):

```php
class Post extends Model {

    /**
     * Relation to (optional) author model
     */
    public function author(): BelongsTo
    {
        return $this->belongsTo(User::class, 'author_id');
    }

    ...
}
```

If `$post->author_id === null`, then that post doesn’t have an author.

However, when I reference the relation with `$post->author` or `$post->load('author')`, then Laravel 5.4 is still tries to query the database.  With debugging, I can see the following SQL is executed:

```
select * from `users` where `users`.`user_id` in ('') and `users`.`deleted_at` is null
```

It seems to me that all the relation logic that gets executed -- and especially the SQL call -- could be skipped if the foreign key to a relation is `null`, no?  Basically, if `author_id` is `null`, then just return `null`.

[One comment](https://github.com/laravel/framework/issues/20616#issuecomment-323509909) suggested that `null` might be a valid foreign key.  While true, I seriously doubt anyone is using `null` as a valid key ... and I might go so far as to suggest that this is an acceptable edge-case to break going forward, if the proposal is accepted.

**Comment by willrowe** [comment]
I believe this is similar to [a PR I submitted last year](https://github.com/laravel/framework/pull/14962) that was rejected.
It might be worth taking a look over that and the related issues I linked to there.

**Comment by cviebrock** [comment]
Thanks @willrowe.  I took a look through your PR and the previous 3 similar issues raised regarding this.

It looks like the consensus is "the relation returns `null` already, so it's not a bug", which I think misses the point of trying to save executing the query in the first place.  Either that, or the issue was closed with very little comment.

The only other objection seems to be the one I mentioned above: someone might legitimately be using `null` as a foreign key.  I see this as unlikely, and a bad practice anyway, so breaking BC for them wouldn't trouble me too much.

Another issue that someone raised (which I can't confirm) is that the current behaviour is actually a bug.  If the foreign model's column is nullable, in which case you won't get an empty result but instead a random element from that table (see https://github.com/laravel/framework/issues/14453#issuecomment-234903691).  I'm not sure if that issue is legit still, or has been solved, but I don't think it's something that would affect what you and I and others have proposed.

We are trying to optimize our Laravel applications as much as possible, and removing unnecessary SQL queries is one of those optimizations.  This proposal would help with that.

**Comment by willrowe** [comment]
> It looks like the consensus is "the relation returns null already, so it's not a bug", which I think misses the point of trying to save executing the query in the first place.

I agree, I created that PR for the same reasons as you. It seems that there isn't much of a desire to make it more performant because of the other drawbacks you mentioned though. A decision would need to be made to either disallow `null` foreign key values or make it configurable.

**Comment by tomgrohl** [comment]
The 'random' case where an element is return is caused by a feature in mysql not a bug in Laravel. (Details here: https://bugs.mysql.com/bug.php?id=62402)

Essentially if `sql_auto_is_null` is on in MySql doing a select against an auto increment column with null will return the last inserted row.

e.g:

```INSERT INTO `t1` (`name`) VALUES ('name1');```

```
SELECT COUNT(*) FROM `t1` WHERE `t1`.`id` IS NULL;
+----------+
| COUNT(*) |
+----------+
|        1 |
+----------+
```

Having `sql_auto_is_null` off fixes this issue. Which is the default.

This caused the issues I was having anyway.

**Comment by crishoj** [comment]
While enabling `sql_auto_is_null` introduces problems in and of itself, it is not the essence of the issue at hand.

Note that the relation might very well be defined on a column that is _not_ auto-incremented.

**Comment by tomgrohl** [comment]
I would have thought editing the model using the relationship would be quite easy. Something like this:

```php

    /**
     * @return CustomerAccount|null
     */
    public function getCustomer()
    {
        if ( ! $this->getAttribute('customer_id')) {
            return null;
        }

        return $this->getAttribute('customer');
    }
    
	/**
     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
     */
    public function customer()
    {
        return $this->belongsTo(CustomerAccount::class, 'customer_id', 'id', 'customer');
    }
```

Making a change to the framework, while it could solve an issue for some people, it might not for others.

**Comment by crishoj** [comment]
Certainly the workaround suggested in https://github.com/laravel/internals/issues/748#issuecomment-328440010 is workable, but my point is that this should be the default behaviour: If the foreign key is missing, it means that there is no related model, and thus no need to run a query looking for a related model with key `NULL`.

**Comment by srottem** [comment]
Given that the database is a scarce resource reducing unnecessary queries is a pretty high priority for me and many others.  I would think that having an default that doesn't execute the query if the foreign key is null and providing a configuration option on the BelongsTo relationship to enable it would be the right way to go.

**Comment by taylorotwell** [comment]
Open to PRs on this issue. It's more of a bug report than an idea.

**Comment by cviebrock** [comment]
Thanks @taylorotwell.  If you consider it a bug, then maybe consider reopening https://github.com/laravel/framework/issues/20616 ?  I'd be happy to work up a PR for it (or @willrowe can resubmit his PR).

**Comment by pionl** [comment]
Hi @cviebrock, do you know if it was fixed in 5.7? In 5.6 is still present. Thanks

**Comment by arthurlauck** [comment]
@pionl Still happening in 5.7
